File: plimsoll/proxy/http_proxy.py
"""
plimsoll.proxy.http_proxy — Web2 Egress Gateway (Python-side).

A Starlette-based HTTP forward proxy that intercepts outgoing API calls
from AI agents and applies Plimsoll CapitalVelocity PID enforcement.

Usage (Python-only deployments)::

    HTTP_PROXY=http://127.0.0.1:8080
    uvicorn plimsoll.proxy.http_proxy:create_http_proxy_app --port 8080

For the full Rust-based TLS proxy, see ``plimsoll-rpc/src/http_proxy.rs``.
"""

from __future__ import annotations

import logging
from typing import Any, Callable, Optional
from urllib.parse import urlparse

from plimsoll.firewall import PlimsollFirewall, PlimsollConfig
from plimsoll.intent import IntentAction, IntentProtocol, NormalizedIntent

logger = logging.getLogger("plimsoll.proxy.http")

# ── Cost extraction rules ─────────────────────────────────────────

API_COST_MAP: dict[str, dict[str, dict[str, Any]]] = {
    "api.stripe.com": {
        "/v1/charges":         {"field": "amount", "divisor": 100.0},
        "/v1/payment_intents": {"field": "amount", "divisor": 100.0},
    },
    "api.openai.com": {
        "/v1/chat/completions": {"fixed_usd": 0.03},
    },
    "api.anthropic.com": {
        "/v1/messages": {"fixed_usd": 0.05},
    },
}


def extract_api_cost(
    domain: str,
    path: str,
    body: Optional[dict[str, Any]] = None,
    cost_map: Optional[dict[str, dict[str, dict[str, Any]]]] = None,
) -> Optional[float]:
    """Extract USD cost from an HTTP request using the cost map.

    Returns ``None`` if the domain/path is not governed.
    """
    cmap = cost_map or API_COST_MAP
    domain_rules = cmap.get(domain)
    if domain_rules is None:
        return None

    for prefix, rule in domain_rules.items():
        if not path.startswith(prefix):
            continue

        # Fixed per-request cost
        if "fixed_usd" in rule:
            return float(rule["fixed_usd"])

        # Extract from body field
        field = rule.get("field")
        if field and body and field in body:
            raw = float(body[field])
            divisor = float(rule.get("divisor", 1.0))
            return raw / divisor

        # Field specified but not in body
        return rule.get("fixed_usd", 0.0)

    return None


def evaluate_http_request(
    firewall: PlimsollFirewall,
    method: str,
    url: str,
    body: Optional[dict[str, Any]] = None,
) -> tuple[bool, str, float]:
    """Evaluate an HTTP API request through the Plimsoll firewall.

    Returns ``(allowed, reason, cost_usd)``.
    """
    parsed = urlparse(url)
    domain = parsed.hostname or ""
    path = parsed.path or "/"

    cost_usd = extract_api_cost(domain, path, body) or 0.0

    intent = NormalizedIntent(
        protocol=IntentProtocol.HTTP,
        action=IntentAction.API_CHARGE,
        capital_at_risk_usd=cost_usd,
        target=url,
        amount_usd=cost_usd,
        function=f"{method.upper()} {url}",
    )

    verdict = firewall.evaluate_intent(intent)
    return verdict.allowed, verdict.reason if verdict.blocked else "allowed", cost_usd

File: plimsoll-rpc/src/http_proxy.rs
//! plimsoll-rpc/src/http_proxy.rs — Web2 Egress Gateway (TLS Interceptor).
//!
//! HTTP/HTTPS Forward Proxy on port `:8080` that intercepts outgoing
//! HTTP requests from AI agents and applies CapitalVelocity PID
//! enforcement to API charges.
//!
//! Enterprises configure `HTTP_PROXY=http://127.0.0.1:8080` in their
//! agent's Docker container.
//!
//! Phase 4.1 of the v2.0 roadmap.

use serde::{Deserialize, Serialize};

// ── Cost rule ────────────────────────────────────────────────────

/// Maps an API endpoint to its USD cost.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiCostRule {
    /// Domain to match (e.g. "api.stripe.com").
    pub domain: String,
    /// Path prefix to match (e.g. "/v1/charges").
    pub path_prefix: String,
    /// Fixed per-request cost in USD (used if `cost_field` is None).
    pub cost_usd: f64,
    /// JSON body field containing the amount (e.g. "amount").
    pub cost_field: Option<String>,
    /// Divisor to convert the field value to USD (e.g. 100 for cents).
    pub cost_divisor: f64,
}

/// Default cost rules for well-known APIs.
pub fn default_cost_rules() -> Vec<ApiCostRule> {
    vec![
        ApiCostRule {
            domain: "api.stripe.com".into(),
            path_prefix: "/v1/charges".into(),
            cost_usd: 0.0,
            cost_field: Some("amount".into()),
            cost_divisor: 100.0, // Stripe uses cents
        },
        ApiCostRule {
            domain: "api.stripe.com".into(),
            path_prefix: "/v1/payment_intents".into(),
            cost_usd: 0.0,
            cost_field: Some("amount".into()),
            cost_divisor: 100.0,
        },
        ApiCostRule {
            domain: "api.openai.com".into(),
            path_prefix: "/v1/chat/completions".into(),
            cost_usd: 0.03, // Approximate per-request cost
            cost_field: None,
            cost_divisor: 1.0,
        },
        ApiCostRule {
            domain: "api.anthropic.com".into(),
            path_prefix: "/v1/messages".into(),
            cost_usd: 0.05, // Approximate per-request cost
            cost_field: None,
            cost_divisor: 1.0,
        },
    ]
}

// ── Cost extraction ──────────────────────────────────────────────

/// Extract the USD cost from an HTTP request.
///
/// Returns `Some(cost_usd)` if a matching rule is found, `None` otherwise.
pub fn extract_cost(
    domain: &str,
    path: &str,
    body: Option<&serde_json::Value>,
    rules: &[ApiCostRule],
) -> Option<f64> {
    for rule in rules {
        if domain != rule.domain || !path.starts_with(&rule.path_prefix) {
            continue;
        }

        // If there's a cost field, extract from body
        if let Some(ref field) = rule.cost_field {
            if let Some(body) = body {
                if let Some(val) = body.get(field) {
                    let amount = val.as_f64().unwrap_or(0.0);
                    return Some(amount / rule.cost_divisor);
                }
            }
            // Field specified but not found in body — return fixed cost
            return Some(rule.cost_usd);
        }

        // No field — use fixed per-request cost
        return Some(rule.cost_usd);
    }

    None // No matching rule — ungoverned domain
}

// ── Analysis result ──────────────────────────────────────────────

/// Result of HTTP request cost analysis.
#[derive(Debug, Clone, Serialize)]
pub struct HttpAnalysisResult {
    pub allowed: bool,
    pub reason: String,
    pub cost_usd: f64,
    pub domain: String,
    pub path: String,
}

// ── Tests ────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stripe_charge_cost_extraction() {
        let rules = default_cost_rules();
        let body = serde_json::json!({ "amount": 5000, "currency": "usd" });
        let cost = extract_cost("api.stripe.com", "/v1/charges", Some(&body), &rules);
        assert_eq!(cost, Some(50.0)); // 5000 cents = $50
    }

    #[test]
    fn test_stripe_payment_intent() {
        let rules = default_cost_rules();
        let body = serde_json::json!({ "amount": 10000 });
        let cost = extract_cost("api.stripe.com", "/v1/payment_intents", Some(&body), &rules);
        assert_eq!(cost, Some(100.0));
    }

    #[test]
    fn test_openai_fixed_cost() {
        let rules = default_cost_rules();
        let body = serde_json::json!({ "model": "gpt-4", "messages": [] });
        let cost = extract_cost("api.openai.com", "/v1/chat/completions", Some(&body), &rules);
        assert_eq!(cost, Some(0.03));
    }

    #[test]
    fn test_ungoverned_domain_returns_none() {
        let rules = default_cost_rules();
        let cost = extract_cost("api.example.com", "/v1/data", None, &rules);
        assert_eq!(cost, None);
    }

    #[test]
    fn test_default_cost_rules_not_empty() {
        let rules = default_cost_rules();
        assert!(rules.len() >= 3);
    }

    #[test]
    fn test_body_field_missing_returns_fixed() {
        let rules = default_cost_rules();
        let body = serde_json::json!({ "currency": "usd" }); // no "amount"
        let cost = extract_cost("api.stripe.com", "/v1/charges", Some(&body), &rules);
        assert_eq!(cost, Some(0.0)); // falls back to cost_usd=0.0
    }
}
